clients/da-vinci-client/src/main/java/com/linkedin/davinci/client/DaVinciRecordTransformer.java

```diff
@@ -34,6 +34,19 @@ public interface DaVinciRecordTransformer<K, V, O extends TransformedRecord> {
    */
   O put(Lazy<K> key, Lazy<V> value);
 
+  /**
+   * By default, calls the put function to be implemented.
+   * @param key to be put
+   * @param value to be put
+   * @param newVersionNumber version number being created
+   * @param oldVersionNumber version number being deleted
+   * @param readyToServe boolean of whether or not PartitionConsumptionState is ready to serve
+   * @return the object to keep in storage, or null if the put should be skipped
+   */
+  default O put(Lazy<K> key, Lazy<V> value, int newVersionNumber, int oldVersionNumber, boolean readyToServe) {
```

Comment:
I realize that we were ambiguous when we discussed "versions" in the last contributor sync. I think what you're going for in this interface is a notion of "version" akin to MVCC or some other row-level versioning mechanism. We do have an equivalent notion to that in Venice which we call "RMD" (Replication Metadata), and it might be interesting to expose that in the interface, though ATM I don't believe we foresee any use case for that...

We also have another notion of versioning, which is the one we're interested in for this interface expansion, and it is much more coarse-grained. In Venice, a dataset is called "store" and each store contains multiple "store-versions". In the current context, a given write operation (put, delete...) will be applied to a single store-version. Or rather, it may be applied to all existing store-versions, but that will occur via separate invokations of this interface. So in this interface, we would want just a single version number param.

Passing the `readyToServe` param is interesting, and might be a viable design. Alternatively, the state change of becoming ready to serve might be best modeled by having a completely separate function. I guess it'd be good to get @ZacAttack's thoughts on that part...

Reply:
readyToServe is interesting.  But....  I'm struggling to think through how it'll be implemented.  I 'think' ready to serve right now is a construct that's per partition, and I don't think it would be a good thing to open up to the user the notion that partition 'X' is ready but parittion 'Y' is not.  I think I'd prefer them to view the whole dataset as ready or not.  So maybe let's remove this from the interface and instead rely on a discrete callback for the lifecycle of a version.